### PR TITLE
*: Support display the raft status of the node in nodes.conditions. #284

### PR DESCRIPTION
**Pull request by runkecheng**
### What type of PR is this?

/enhancement

### Which issue(s) this PR fixes?

Fixes ##284

### What this PR does?

Summary:

- Adding the basic framework code of the xenon http request.
- Support `raft/status` and `raft/trytoleader` 
- Support display the raft status of the node in nodes.conditions.

### CODE REVIEW

**Comment by runkecheng** [comment]
How to take a short little alias for `internal`.
```
/home/runner/work/radondb-mysql-kubernetes/radondb-mysql-kubernetes/mysqlcluster/syncer/statefulset.go
  (41, 2)  ST1001  should not use dot imports
```

**Comment by runkecheng** [comment]
After modification:
```
  Nodes:
    Conditions:
      Last Transition Time:  2021-11-04T09:53:05Z
      Status:                False
      Type:                  Lagged
      Last Transition Time:  2021-11-04T09:53:05Z
      Status:                True
      Type:                  Leader
      Last Transition Time:  2021-11-04T09:53:05Z
      Status:                False
      Type:                  ReadOnly
      Last Transition Time:  2021-11-04T09:53:05Z
      Status:                False
      Type:                  Replicating
    Name:                    sample-mysql-0.sample-mysql.default
    Raft Status:
      Leader:  sample-mysql-0.sample-mysql.default:8801
      Nodes:
        sample-mysql-0.sample-mysql.default:8801
        sample-mysql-1.sample-mysql.default:8801
        sample-mysql-2.sample-mysql.default:8801
      Role:  LEADER
    Conditions:
      Last Transition Time:  2021-11-04T09:55:25Z
      Status:                False
      Type:                  Lagged
      Last Transition Time:  2021-11-04T09:55:25Z
      Status:                False
      Type:                  Leader
      Last Transition Time:  2021-11-04T09:55:25Z
      Status:                True
      Type:                  ReadOnly
      Last Transition Time:  2021-11-04T09:55:25Z
      Status:                True
      Type:                  Replicating
    Name:                    sample-mysql-1.sample-mysql.default
    Raft Status:
      Leader:  sample-mysql-0.sample-mysql.default:8801
      Nodes:
        sample-mysql-0.sample-mysql.default:8801
        sample-mysql-1.sample-mysql.default:8801
        sample-mysql-2.sample-mysql.default:8801
      Role:  FOLLOWER
    Conditions:
      Last Transition Time:  2021-11-04T09:58:25Z
      Status:                False
      Type:                  Lagged
      Last Transition Time:  2021-11-04T09:58:25Z
      Status:                False
      Type:                  Leader
      Last Transition Time:  2021-11-04T09:58:25Z
      Status:                True
      Type:                  ReadOnly
      Last Transition Time:  2021-11-04T09:58:25Z
      Status:                True
      Type:                  Replicating
    Name:                    sample-mysql-2.sample-mysql.default
    Raft Status:
      Leader:  sample-mysql-0.sample-mysql.default:8801
      Nodes:
        sample-mysql-0.sample-mysql.default:8801
        sample-mysql-1.sample-mysql.default:8801
        sample-mysql-2.sample-mysql.default:8801
      Role:     FOLLOWER
```

File: internal/http_executor.go [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+	client *http.Client
+}
+
+func NewHttpClient(client *http.Client) HttpClient {
+	return &httpClient{
+		client: client,
+	}
+}
+
+func (httpClient *httpClient) Do(req *http.Request) (*http.Response, error) {
```
Comment (acekingke) [inline internal/http_executor.go:47]: httpClient variable is same to type， use `client` may be enough?

File: internal/xenon_request.go [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+// NewxenononnhttpRequest returns the http request of the corresponding type according to the incoming configuration.
+func NewXenonHttpRequest(cfg RequestConfig) (*Request, error) {
+	requestType, ok := utils.XenonHttpUrls[string(cfg.xenonHttpUrl)]
+	if !ok {
+		return nil, fmt.Errorf("xenon http url[%s] does not exist or does not support", cfg.xenonHttpUrl)
+	}
+
+	switch requestType {
+	case http.MethodGet:
+		req, err := newHttpGetRequest(cfg)
```
Comment (acekingke) [inline internal/xenon_request.go:54]: `return newHttpGetRequest(cfg) ` is enough.

File: internal/xenon_request.go [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+		if err != nil {
+			return nil, err
+		}
+		return req, nil
+	case http.MethodPost:
+		req, err := newHttpPostRequest(cfg)
```
Comment (acekingke) [inline internal/xenon_request.go:60]: The same to above

File: mysqlcluster/syncer/statefulset.go [hunk 4]
```diff
@@ -66,10 +63,12 @@ type StatefulSetSyncer struct {
 
 	// Mysql query runner.
 	internal.SQLRunnerFactory
+
+	internal.XenonExecutor
```
Comment (acekingke) [inline mysqlcluster/syncer/statefulset.go:67]: add comment like above?

File: mysqlcluster/syncer/statefulset.go [hunk 5]
```diff
@@ -422,18 +422,14 @@ func (s *StatefulSetSyncer) preUpdate(ctx context.Context, leader, follower stri
...
 	if err = retry(time.Second*5, time.Second*60, func() (bool, error) {
-		// Check whether is leader.
-		status, err := checkRole(followerHost, rootPasswd)
-		if err != nil {
-			log.Error(err, "failed to check role", "pod", follower)
-			return false, nil
-		}
-		if status == corev1.ConditionTrue {
-			return true, nil
+		for _, node := range s.Status.Nodes {
```
Comment (acekingke) [inline mysqlcluster/syncer/statefulset.go:425]: this code has bug， the `s.Status.Nodes` never change

File: internal/xenon_executor.go [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+func (xenonExecutor *xenonExecutor) GetRootPassword() string {
+	return xenonExecutor.rootPassword
+}
+
+func (xenonExecutor *xenonExecutor) SetRootPassword(rootPassword string) {
+	xenonExecutor.rootPassword = rootPassword
+}
+
+// Raftstatus gets the raft status of incoming host through http.
+func (xenonExecutor *xenonExecutor) RaftStatus(host string) (*apiv1alpha1.RaftStatus, error) {
```
Comment (acekingke) [inline internal/xenon_executor.go:52]: variables same with type

**Comment by acekingke** [comment]
LGTM

File: internal/xenon_request.go [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+func NewRequestConfig(host, rootPasswd string, xenonHttpUrl utils.XenonHttpUrl, data interface{}) RequestConfig {
+	return RequestConfig{
+		rootPassword: rootPasswd,
+		host:         host,
+		xenonHttpUrl: xenonHttpUrl,
+		data:         data,
+	}
+}
+
+// NewxenononnhttpRequest returns the http request of the corresponding type according to the incoming configuration.
```
Comment (andyli029) [inline internal/xenon_request.go:45]: Fix the line.

**Comment by andyli029** [comment]
LGTM.

**Pull request merged**